Add filter to allow overriding all product attributes

nima-karimi opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

This PR introduces a new filter called `woocommerce_gla_product_attribute_values` to allow overriding all product properties.

The filter accepts a list of key => values containing Google product properties. Basically, any property of the `\Google\Service\ShoppingContent\Product` class can be overridden and set using this filter now. This allows for more customization in cases where our mapping is not working as expected for users.


### Detailed test instructions:

1. Use the following filter configuration to provide a test value for `description`:
```
add_filter(
  'woocommerce_gla_product_attribute_values',
  function ( array $attributes, WC_Product $product, WCProductAdapter $google_product ) {
    $attributes['description'] = 'Overridden description! Previously set to: ' . $product->get_description();

    return $attributes;
  },
  10,
  3
);
```
2. Sync a product and make sure the description of the synced product in MC is now set to the value provided by the filter

Note: We already provide filters for setting custom attribute values like `GTIN`, `MPN`, etc. However, those filters do now allow setting basic attributes such as image link or description. My aim was to allow overriding any and all attributes when mapping a WooCommerce product so that third parties have more flexibility for integration. Not sure if we should remove those now that we have this filter because this one overlaps with those filters and overrides any value set by those filters.

### Changelog entry

> Add `woocommerce_gla_product_attribute_values` filter to allow overriding all product attributes.

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted. 
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Or leave the "Changelog entry" header in place without any summary if no changelog entry is needed.  
Otherwise, the title of Pull Request will be used as the changelog entry.  
-->

nima-karimi committed:
Introduces a new filter to allow overriding all product properties

The filter is called `woocommerce_gla_override_product_attribute_values` and accepts a list of key => values containing Google product properties. Basically any property of the `\Google\Service\ShoppingContent\Product` class can be overridden and set using this filter now. This allows for more customization in cases where our mapping is not working as expected for users.

nima-karimi committed:
Add unit tests for overriding basic properties via filter

mikkamp reviewed (approved):
Code looks good and is working as intended. I personally wouldn't have included `override` in the filter name. Since one of the main purposes of a filter is to allow you to override something.

Regarding multiple filters to override the same thing, I think we should leave that as is. It's common to have both specific filters and a larger one for all the data, in WooCommerce and other extensions. If we need to clarify which filter takes preference, we can do so through a code comment or in the documentation.

nima-karimi committed:
Add documentation to attribute override filters

indicating which one takes precedence and which attributes do they modify

nima-karimi committed:
rename filter

nima-karimi committed:
fix doc reference

nima-karimi committed:
add more unit tests for override filter

nima-karimi commented:
> I personally wouldn't have included override in the filter name. Since one of the main purposes of a filter is to allow you to override something.

Ah yeah, I had some doubts there too. I've removed the `override` from the naming in https://github.com/woocommerce/google-listings-and-ads/pull/925/commits/3e8a593da77776398a4f14ba1f7131cf23362e2b

> If we need to clarify which filter takes preference, we can do so through a code comment or in the documentation.

Added some more documentation in the code to specify what each filter does and how they work together.
